source/common/router/config_impl.cc

```diff
@@ -896,32 +900,36 @@ RouteConstSharedPtr PathRouteEntryImpl::matches(const Http::HeaderMap& headers,
 RegexRouteEntryImpl::RegexRouteEntryImpl(const VirtualHostImpl& vhost,
                                          const envoy::api::v2::route::Route& route,
                                          Server::Configuration::FactoryContext& factory_context)
-    : RouteEntryImplBase(vhost, route, factory_context),
-      regex_(RegexUtil::parseRegex(route.match().regex())), regex_str_(route.match().regex()) {}
+    : RouteEntryImplBase(vhost, route, factory_context) {
+  if (route.match().path_specifier_case() == envoy::api::v2::route::RouteMatch::kRegex) {
+    regex_ = Regex::Utility::parseStdRegexAsCompiledMatcher(route.match().regex());
+    regex_str_ = route.match().regex();
+  } else {
+    ASSERT(route.match().path_specifier_case() == envoy::api::v2::route::RouteMatch::kSafeRegex);
+    regex_ = Regex::Utility::parseRegex(route.match().safe_regex());
```

Comment:
My main observation is that there are a few of these patterns throughout the code where we have separate entries for regex and safe-regex, and presumably there are also some other ones nearby for prefix/suffix matcher. And there's a lot of repeated conditional patterns throughout the code.

What I was wondering was whether this might be an opportunity to try to have a design where the condiitonalization was factored out. You could have a Pattern or Matcher class, which in every application we'd have something like:

  matcher_ = std::make_unique<Matcher>(pattern, type)

where pattern and type came from the config proto., and type is one of Regex, SafeRegex, PrefixMatch, SuffixMatch, ExactMatch, Wildcard, etc.

The Matcher might have some helper classes behind the scenes where we have an ABC & 5 impls. 

My thinking (not having gotten too deep into this part of the codebase yet) is that this might factor out some of the complexity, and possibly pave the way to doing more interesting things with scalable routing structures.


Moreover, I feel like a MatcherFactory might be a good abstraction as it would give us a place to conveniently store a proxy-wide ProgramSize limit, and maybe some counters about how much regex evaluation we are actually doing, which might be interesting when we're doing data-plane perf analysis with large tables.

Reply:
Unless I am misunderstanding, what you are asking for is effectively what `StringMatcher` is. The issue is that we have a lot of tech debt over the last couple of years as we have moved towards unified matchers. Per the comment from @lizan above, I am going to switch `VirtualCluster` to use a string matcher, but IMO it's too disruptive to change route matching and header matching to use string matcher, as this will break a huge number of users. My personal feeling is that we should clean this up in the v3 API sweep to make sure that everything is using common matchers that should be. How does that sound?

Reply:
sgtm & understood.

Reply:
@mattklein123 isn't `StringMatcher` strictly more powerful than `RegexMatcher`? If this is a new field we are adding, why does that break existing users if we shoot for the more flexible one?

Reply:
The route matching code is special. Regex for example strips query path, etc. I suppose we could do some type of route matcher that is a string matcher with query path stripped or something like that. I will think about it.

Reply:
Please take a look at the comment I just added in route.proto. I've thought about this and IMO its not worth it to boil the ocean here in the v2 API. I think we should fix and rethink this in the v3 API. Let me know what you think after you read the comment.